drivers/gpio/gpio_bcm2711.c

```diff
@@ -0,0 +1,303 @@
+	if (!status) {
+		/* Nothing happened in this instance */
+		return;
+	}
+
+	gpio_fire_callbacks(&data->cb, port, status);
+
+	/* Write to clear */
+	sys_write32(status, GPEDS(data->base, cfg->idx));
+}
+
+int gpio_bcm2711_init(const struct device *port)
+{
+	struct gpio_bcm2711_data *data = DEV_DATA(port);
+
+	DEVICE_MMIO_NAMED_MAP(port, reg_base, K_MEM_CACHE_NONE);
+	data->base = DEVICE_MMIO_NAMED_GET(port, reg_base);
+
+	return 0;
+}
+
+static const struct gpio_driver_api gpio_bcm2711_api = {
+	.pin_configure = gpio_bcm2711_pin_configure,
+	.port_get_raw = gpio_bcm2711_port_get_raw,
+	.port_set_masked_raw = gpio_bcm2711_port_set_masked_raw,
+	.port_set_bits_raw = gpio_bcm2711_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_bcm2711_port_clear_bits_raw,
+	.port_toggle_bits = gpio_bcm2711_port_toggle_bits,
+	.pin_interrupt_configure = gpio_bcm2711_pin_interrupt_configure,
+	.manage_callback = gpio_bcm2711_manage_callback,
+};
+
+#define GPIO_BCM2711_INST(n)                                                                       \
+	static struct gpio_bcm2711_data gpio_bcm2711_data_##n;                                     \
+                                                                                                   \
+	static const struct gpio_bcm2711_config gpio_bcm2711_cfg_##n = {                           \
+		.common = {.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(0)},                   \
+		DEVICE_MMIO_NAMED_ROM_INIT(reg_base, DT_INST_PARENT(n)),                           \
+		.idx = DT_INST_REG_ADDR(n),                                                        \
+		.isr = gpio_bcm2711_isr,                                                           \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(n, gpio_bcm2711_init, NULL, &gpio_bcm2711_data_##n,                  \
+			      &gpio_bcm2711_cfg_##n, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,      \
+			      &gpio_bcm2711_api);
+
+DT_INST_FOREACH_STATUS_OKAY(GPIO_BCM2711_INST)
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+#define GPIO_BCM2711_CONTROLLER DT_PARENT(DT_COMPAT_GET_ANY_STATUS_OKAY(DT_DRV_COMPAT))
+
+#define GPIO_BCM2711_ISR_DISPATCH(n)                                                               \
+	const struct device *port_##n = DEVICE_DT_INST_GET(n);                                     \
+	DEV_CFG(port_##n)->isr(port_##n);
+
+static void uart_bcm2711_controller_isr(void *p)
+{
+	ARG_UNUSED(p);
+	DT_INST_FOREACH_STATUS_OKAY(GPIO_BCM2711_ISR_DISPATCH);
+}
+
+static int uart_bcm2711_controller_init(void)
+{
+	IRQ_CONNECT(DT_IRQN(GPIO_BCM2711_CONTROLLER), DT_IRQ(GPIO_BCM2711_CONTROLLER, priority),
+		    uart_bcm2711_controller_isr, NULL, 0);
+	irq_enable(DT_IRQN(GPIO_BCM2711_CONTROLLER));
+
+	return 0;
+}
+
+SYS_INIT(uart_bcm2711_controller_init, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY);
```

Comment:
If I read this correctly, SYS_INIT is not required, just define this function as part of the device definition macro and call it during driver init, there are quite a few examples in tree

Reply:
It's strange that shared IRQ causes the applicaton hangs "*** Booti". I'm investigating the problem.

Reply:
Fixed